edge-util/src/Microsoft.Azure.Devices.Edge.Util/TaskEx.cs

```diff
@@ -286,5 +288,52 @@ static async Task ExecuteUntilCancelled(this Task task, CancellationToken cancel
             //// Await here to bubble up any exceptions
             await completedTask;
         }
+
+        public static async Task<IEnumerable<TOut>> SelectAsync<TIn, TOut>(this IEnumerable<TIn> source, Func<TIn, Task<TOut>> func, int concurrentLimit = int.MaxValue)
+        {
+            // If number of items to be processed is less than limit, just run all
+            if (source.Count() <= concurrentLimit)
+            {
+                return await Task.WhenAll(source.Select(func));
+            }
+
+            // Lock for enumerator and result list since they are accessed by independent tasks
+            AsyncLock enumerator_lock = new AsyncLock();
+            List<TOut> result = new List<TOut>();
+            var enumerator = source.GetEnumerator();
+
+            // recursive thread safe function that will drain [source] and call [func] on each element until source is empty
+            async Task ProcessNext()
+            {
+                TIn current;
+                using (await enumerator_lock.LockAsync())
+                {
+                    if (!enumerator.MoveNext())
+                    {
+                        // If enumerator is empty, end recursion
+                        // Note lock is released by using statement
+                        return;
+                    }
+
+                    current = enumerator.Current;
+                }
+
+                // Do work
+                TOut value = await func(current);
+
+                // Append result to result list
+                using (await enumerator_lock.LockAsync())
+                {
+                    result.Add(value);
```

Comment:
is ordering important? this extension can be used in many places, I guess. It is worth to mention the ordering guarantees (or their absence in this case), or fix the ordering.

To fix the ordering you can pass an index for every task and collect the results in some (ordered?) Dictionary with index as a key.

Reply:
Ordering isn't important, all that matters is that all tasks are completed. If I was worried about ordering, I would just use an array since I know the final size.

Reply:
Ok, let's just put some comments on a method stating that, so no surprises in the future :)